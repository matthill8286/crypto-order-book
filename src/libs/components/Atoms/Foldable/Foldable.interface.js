
//# sourceMappingURL=Foldable.interface.js.map