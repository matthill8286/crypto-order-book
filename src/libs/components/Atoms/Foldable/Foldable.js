function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import { styled } from "../../../styles/styled";
export var StyledFoldable = styled.div.withConfig({
  displayName: "Foldable__StyledFoldable",
  componentId: "sc-1ors3ot-0"
})(["overflow:hidden;", ""], function (_ref) {
  var looksOpenFromBreakpoint = _ref.looksOpenFromBreakpoint;
  return looksOpenFromBreakpoint && "\n    @media (min-width: ".concat(looksOpenFromBreakpoint, "px) {\n      height: auto !important;\n    }\n  ");
});
export var Foldable = function (_React$Component) {
  _inherits(Foldable, _React$Component);

  var _super = _createSuper(Foldable);

  function Foldable(props) {
    var _this;

    _classCallCheck(this, Foldable);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "entryRef", void 0);

    _defineProperty(_assertThisInitialized(_this), "standardDuration", void 0);

    _defineProperty(_assertThisInitialized(_this), "maxDuration", void 0);

    _defineProperty(_assertThisInitialized(_this), "_afterTransition", function () {
      var isOpen = _this.props.isOpen;

      if (isOpen) {
        _this.setState({
          height: 'auto'
        });
      }
    });

    _this.entryRef = React.createRef();
    _this.standardDuration = 300;
    _this.maxDuration = 1000;
    _this.state = {
      height: _this._getHeight(),
      transition: "height ".concat(_this.standardDuration, "ms linear"),
      looksOpenFromBreakpoint: props.looksOpenInitiallyFromBreakpoint
    };
    return _this;
  }

  _createClass(Foldable, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      if (this.props.isOpen) {
        this._updateHeightBeforeTransition();
      }
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (this.props.isOpen !== prevProps.isOpen) {
        this._updateHeightBeforeTransition();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          _isOpen = _this$props.isOpen,
          rest = _objectWithoutProperties(_this$props, ["isOpen"]);

      var looksOpenFromBreakpoint = this.state.looksOpenFromBreakpoint;
      return React.createElement(StyledFoldable, _extends({
        "data-test": 'foldable',
        style: this._applyTransition(),
        onTransitionEnd: this._afterTransition,
        className: this.props.className,
        looksOpenFromBreakpoint: looksOpenFromBreakpoint
      }, rest), React.createElement("div", {
        "data-test": "inner-foldable",
        ref: this.entryRef
      }, this.props.children));
    }
  }, {
    key: "_scrollToElement",
    value: function _scrollToElement(duration) {
      var currentElement = this.entryRef.current;
      setTimeout(function () {
        var rect = (currentElement === null || currentElement === void 0 ? void 0 : currentElement.getBoundingClientRect()) || {
          top: 0
        };

        if (rect.top < 0) {
          window.scrollTo(0, 0);
          var rectUpdate = (currentElement === null || currentElement === void 0 ? void 0 : currentElement.getBoundingClientRect()) || {
            top: 0,
            height: 0
          };
          window.scrollTo(0, rectUpdate.top - rectUpdate.height);
        }
      }, duration);
    }
  }, {
    key: "_updateHeightBeforeTransition",
    value: function _updateHeightBeforeTransition() {
      var _currentElement$clien,
          _this2 = this;

      var looksOpenFromBreakpoint = this.state.looksOpenFromBreakpoint;
      var isTransitionCancelled = !!looksOpenFromBreakpoint;
      var currentElement = this.entryRef.current;

      var calculatedDuration = this._getDuration();

      var animationType = this._getAnimationType(calculatedDuration);

      var height = isTransitionCancelled ? 'auto' : (_currentElement$clien = currentElement === null || currentElement === void 0 ? void 0 : currentElement.clientHeight) !== null && _currentElement$clien !== void 0 ? _currentElement$clien : 'auto';
      var transition = isTransitionCancelled ? '' : "height ".concat(calculatedDuration, "ms ").concat(animationType);
      this.setState({
        height: height,
        transition: transition,
        looksOpenFromBreakpoint: undefined
      }, function () {
        if (!_this2.props.isOpen) {
          _this2._updateTransition();
        } else {
          _this2._scrollToElement(calculatedDuration);
        }
      });
    }
  }, {
    key: "_getDuration",
    value: function _getDuration() {
      var isOpen = this.props.isOpen;
      var currentElement = this.entryRef.current;
      var calcDuration = !!currentElement && currentElement.clientHeight / 2;

      if (calcDuration) {
        if (!isOpen) {
          return this.standardDuration;
        } else if (calcDuration > this.maxDuration) {
          return this.maxDuration;
        } else if (calcDuration < this.standardDuration) {
          return this.standardDuration;
        } else {
          return calcDuration;
        }
      } else {
        return this.standardDuration;
      }
    }
  }, {
    key: "_getAnimationType",
    value: function _getAnimationType(calculatedDuration) {
      var isOpen = this.props.isOpen;
      return calculatedDuration === this.standardDuration ? 'linear' : isOpen ? 'ease-in' : 'ease-out';
    }
  }, {
    key: "_updateTransition",
    value: function _updateTransition() {
      var calculatedDuration = this._getDuration();

      var animationType = this._getAnimationType(calculatedDuration);

      this.setState({
        height: this._getHeight(),
        transition: "height ".concat(calculatedDuration, "ms ").concat(animationType)
      });
    }
  }, {
    key: "_applyTransition",
    value: function _applyTransition() {
      return {
        height: this.state.height,
        transition: this.state.transition
      };
    }
  }, {
    key: "_getHeight",
    value: function _getHeight() {
      var isOpen = this.props.isOpen;
      return isOpen ? 'auto' : 0;
    }
  }]);

  return Foldable;
}(React.Component);
//# sourceMappingURL=Foldable.js.map