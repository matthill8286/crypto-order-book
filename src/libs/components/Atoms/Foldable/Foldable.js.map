{"version":3,"sources":["../../../../src/components/Atoms/Foldable/Foldable.tsx"],"names":["React","styled","StyledFoldable","div","looksOpenFromBreakpoint","Foldable","props","isOpen","setState","height","entryRef","createRef","standardDuration","maxDuration","state","_getHeight","transition","looksOpenInitiallyFromBreakpoint","_updateHeightBeforeTransition","prevProps","_isOpen","rest","_applyTransition","_afterTransition","className","children","duration","currentElement","current","setTimeout","rect","getBoundingClientRect","top","window","scrollTo","rectUpdate","isTransitionCancelled","calculatedDuration","_getDuration","animationType","_getAnimationType","clientHeight","undefined","_updateTransition","_scrollToElement","calcDuration","Component"],"mappings":";;;;;;;;;;;;;;;;;;;;;;;;;;;;;;AAAA,OAAOA,KAAP,MAAkB,OAAlB;AACA,SAASC,MAAT;AAGA,OAAO,IAAMC,cAAc,GAAGD,MAAM,CAACE,GAAV;AAAA;AAAA;AAAA,6BAGvB;AAAA,MAAGC,uBAAH,QAAGA,uBAAH;AAAA,SACAA,uBAAuB,uCAEFA,uBAFE,qDADvB;AAAA,CAHuB,CAApB;AAYP,WAAaC,QAAb;AAAA;;AAAA;;AAKE,oBAAYC,KAAZ,EAAkC;AAAA;;AAAA;;AAChC,8BAAMA,KAAN;;AADgC;;AAAA;;AAAA;;AAAA,uEA2CP,YAAM;AAAA,UACvBC,MADuB,GACZ,MAAKD,KADO,CACvBC,MADuB;;AAE/B,UAAIA,MAAJ,EAAY;AACV,cAAKC,QAAL,CAAc;AACZC,UAAAA,MAAM,EAAE;AADI,SAAd;AAGD;AACF,KAlDiC;;AAEhC,UAAKC,QAAL,GAAgBV,KAAK,CAACW,SAAN,EAAhB;AACA,UAAKC,gBAAL,GAAwB,GAAxB;AACA,UAAKC,WAAL,GAAmB,IAAnB;AAEA,UAAKC,KAAL,GAAa;AACXL,MAAAA,MAAM,EAAE,MAAKM,UAAL,EADG;AAEXC,MAAAA,UAAU,mBAAY,MAAKJ,gBAAjB,cAFC;AAGXR,MAAAA,uBAAuB,EAAEE,KAAK,CAACW;AAHpB,KAAb;AANgC;AAWjC;;AAhBH;AAAA;AAAA,WAkBE,6BAA2B;AACzB,UAAI,KAAKX,KAAL,CAAWC,MAAf,EAAuB;AACrB,aAAKW,6BAAL;AACD;AACF;AAtBH;AAAA;AAAA,WAwBE,4BAA0BC,SAA1B,EAAoD;AAClD,UAAI,KAAKb,KAAL,CAAWC,MAAX,KAAsBY,SAAS,CAACZ,MAApC,EAA4C;AAC1C,aAAKW,6BAAL;AACD;AACF;AA5BH;AAAA;AAAA,WA8BE,kBAAgB;AAAA,wBACuB,KAAKZ,KAD5B;AAAA,UACEc,OADF,eACNb,MADM;AAAA,UACcc,IADd;;AAAA,UAENjB,uBAFM,GAEsB,KAAKU,KAF3B,CAENV,uBAFM;AAGd,aACE,oBAAC,cAAD;AACE,qBAAW,UADb;AAEE,QAAA,KAAK,EAAE,KAAKkB,gBAAL,EAFT;AAGE,QAAA,eAAe,EAAE,KAAKC,gBAHxB;AAIE,QAAA,SAAS,EAAE,KAAKjB,KAAL,CAAWkB,SAJxB;AAKE,QAAA,uBAAuB,EAAEpB;AAL3B,SAMMiB,IANN,GAOE;AAAK,qBAAU,gBAAf;AAAgC,QAAA,GAAG,EAAE,KAAKX;AAA1C,SACG,KAAKJ,KAAL,CAAWmB,QADd,CAPF,CADF;AAaD;AA9CH;AAAA;AAAA,WAyDE,0BAAyBC,QAAzB,EAA2C;AACzC,UAAMC,cAAc,GAAG,KAAKjB,QAAL,CAAckB,OAArC;AAEAC,MAAAA,UAAU,CAAC,YAAM;AACf,YAAMC,IAAI,GAAG,CAAAH,cAAc,SAAd,IAAAA,cAAc,WAAd,YAAAA,cAAc,CAAEI,qBAAhB,OAA2C;AAAEC,UAAAA,GAAG,EAAE;AAAP,SAAxD;;AACA,YAAIF,IAAI,CAACE,GAAL,GAAW,CAAf,EAAkB;AAEhBC,UAAAA,MAAM,CAACC,QAAP,CAAgB,CAAhB,EAAmB,CAAnB;AACA,cAAMC,UAAU,GAAG,CAAAR,cAAc,SAAd,IAAAA,cAAc,WAAd,YAAAA,cAAc,CAAEI,qBAAhB,OAA2C;AAAEC,YAAAA,GAAG,EAAE,CAAP;AAAUvB,YAAAA,MAAM,EAAE;AAAlB,WAA9D;AACAwB,UAAAA,MAAM,CAACC,QAAP,CAAgB,CAAhB,EAAmBC,UAAU,CAACH,GAAX,GAAiBG,UAAU,CAAC1B,MAA/C;AACD;AACF,OARS,EAQPiB,QARO,CAAV;AASD;AArEH;AAAA;AAAA,WAuEE,yCAAwC;AAAA;AAAA;;AAAA,UAC9BtB,uBAD8B,GACF,KAAKU,KADH,CAC9BV,uBAD8B;AAEtC,UAAMgC,qBAAqB,GAAG,CAAC,CAAChC,uBAAhC;AACA,UAAMuB,cAAc,GAAG,KAAKjB,QAAL,CAAckB,OAArC;;AACA,UAAMS,kBAAkB,GAAG,KAAKC,YAAL,EAA3B;;AACA,UAAMC,aAAa,GAAG,KAAKC,iBAAL,CAAuBH,kBAAvB,CAAtB;;AAEA,UAAM5B,MAAM,GAAG2B,qBAAqB,GAAG,MAAH,4BAAYT,cAAZ,aAAYA,cAAZ,uBAAYA,cAAc,CAAEc,YAA5B,yEAA4C,MAAhF;AACA,UAAMzB,UAAU,GAAGoB,qBAAqB,GACpC,EADoC,oBAE1BC,kBAF0B,gBAEFE,aAFE,CAAxC;AAIA,WAAK/B,QAAL,CACE;AACEC,QAAAA,MAAM,EAANA,MADF;AAEEO,QAAAA,UAAU,EAAVA,UAFF;AAGEZ,QAAAA,uBAAuB,EAAEsC;AAH3B,OADF,EAME,YAAM;AACJ,YAAI,CAAC,MAAI,CAACpC,KAAL,CAAWC,MAAhB,EAAwB;AACtB,UAAA,MAAI,CAACoC,iBAAL;AACD,SAFD,MAEO;AACL,UAAA,MAAI,CAACC,gBAAL,CAAsBP,kBAAtB;AACD;AACF,OAZH;AAcD;AAjGH;AAAA;AAAA,WAmGE,wBAAuB;AAAA,UACb9B,MADa,GACF,KAAKD,KADH,CACbC,MADa;AAErB,UAAMoB,cAAc,GAAG,KAAKjB,QAAL,CAAckB,OAArC;AACA,UAAMiB,YAAY,GAAG,CAAC,CAAClB,cAAF,IAAoBA,cAAc,CAACc,YAAf,GAA8B,CAAvE;;AAEA,UAAII,YAAJ,EAAkB;AAChB,YAAI,CAACtC,MAAL,EAAa;AACX,iBAAO,KAAKK,gBAAZ;AACD,SAFD,MAEO,IAAIiC,YAAY,GAAG,KAAKhC,WAAxB,EAAqC;AAC1C,iBAAO,KAAKA,WAAZ;AACD,SAFM,MAEA,IAAIgC,YAAY,GAAG,KAAKjC,gBAAxB,EAA0C;AAC/C,iBAAO,KAAKA,gBAAZ;AACD,SAFM,MAEA;AACL,iBAAOiC,YAAP;AACD;AACF,OAVD,MAUO;AACL,eAAO,KAAKjC,gBAAZ;AACD;AACF;AArHH;AAAA;AAAA,WAuHE,2BAA0ByB,kBAA1B,EAAsD;AAAA,UAC5C9B,MAD4C,GACjC,KAAKD,KAD4B,CAC5CC,MAD4C;AAEpD,aAAO8B,kBAAkB,KAAK,KAAKzB,gBAA5B,GAA+C,QAA/C,GAA0DL,MAAM,GAAG,SAAH,GAAe,UAAtF;AACD;AA1HH;AAAA;AAAA,WA4HE,6BAA4B;AAC1B,UAAM8B,kBAAkB,GAAG,KAAKC,YAAL,EAA3B;;AACA,UAAMC,aAAa,GAAG,KAAKC,iBAAL,CAAuBH,kBAAvB,CAAtB;;AACA,WAAK7B,QAAL,CAAc;AACZC,QAAAA,MAAM,EAAE,KAAKM,UAAL,EADI;AAEZC,QAAAA,UAAU,mBAAYqB,kBAAZ,gBAAoCE,aAApC;AAFE,OAAd;AAID;AAnIH;AAAA;AAAA,WAqIE,4BAA2B;AACzB,aAAO;AACL9B,QAAAA,MAAM,EAAE,KAAKK,KAAL,CAAWL,MADd;AAELO,QAAAA,UAAU,EAAE,KAAKF,KAAL,CAAWE;AAFlB,OAAP;AAID;AA1IH;AAAA;AAAA,WA4IE,sBAAqB;AAAA,UACXT,MADW,GACA,KAAKD,KADL,CACXC,MADW;AAEnB,aAAOA,MAAM,GAAG,MAAH,GAAY,CAAzB;AACD;AA/IH;;AAAA;AAAA,EAA8BP,KAAK,CAAC8C,SAApC","sourcesContent":["import React from 'react'\nimport { styled } from '@/styles/styled'\nimport { FoldableProps, FoldableState } from './Foldable.interface'\n\nexport const StyledFoldable = styled.div<{ looksOpenFromBreakpoint?: number }>`\n  overflow: hidden;\n\n  ${({ looksOpenFromBreakpoint }) =>\n    looksOpenFromBreakpoint &&\n    `\n    @media (min-width: ${looksOpenFromBreakpoint}px) {\n      height: auto !important;\n    }\n  `}\n`\n\nexport class Foldable extends React.Component<FoldableProps, FoldableState> {\n  private entryRef: React.RefObject<HTMLDivElement>\n  private standardDuration: number\n  private maxDuration: number\n\n  constructor(props: FoldableProps) {\n    super(props)\n    this.entryRef = React.createRef()\n    this.standardDuration = 300\n    this.maxDuration = 1000\n\n    this.state = {\n      height: this._getHeight(),\n      transition: `height ${this.standardDuration}ms linear`,\n      looksOpenFromBreakpoint: props.looksOpenInitiallyFromBreakpoint,\n    }\n  }\n\n  public componentDidMount() {\n    if (this.props.isOpen) {\n      this._updateHeightBeforeTransition()\n    }\n  }\n\n  public componentDidUpdate(prevProps: FoldableProps) {\n    if (this.props.isOpen !== prevProps.isOpen) {\n      this._updateHeightBeforeTransition()\n    }\n  }\n\n  public render() {\n    const { isOpen: _isOpen, ...rest } = this.props\n    const { looksOpenFromBreakpoint } = this.state\n    return (\n      <StyledFoldable\n        data-test={'foldable'}\n        style={this._applyTransition()}\n        onTransitionEnd={this._afterTransition}\n        className={this.props.className}\n        looksOpenFromBreakpoint={looksOpenFromBreakpoint}\n        {...rest}>\n        <div data-test=\"inner-foldable\" ref={this.entryRef}>\n          {this.props.children}\n        </div>\n      </StyledFoldable>\n    )\n  }\n\n  private _afterTransition = () => {\n    const { isOpen } = this.props\n    if (isOpen) {\n      this.setState({\n        height: 'auto',\n      })\n    }\n  }\n\n  private _scrollToElement(duration: number) {\n    const currentElement = this.entryRef.current\n\n    setTimeout(() => {\n      const rect = currentElement?.getBoundingClientRect() || { top: 0 }\n      if (rect.top < 0) {\n        // reset viewport to calculate the right position\n        window.scrollTo(0, 0)\n        const rectUpdate = currentElement?.getBoundingClientRect() || { top: 0, height: 0 }\n        window.scrollTo(0, rectUpdate.top - rectUpdate.height)\n      }\n    }, duration)\n  }\n\n  private _updateHeightBeforeTransition() {\n    const { looksOpenFromBreakpoint } = this.state\n    const isTransitionCancelled = !!looksOpenFromBreakpoint\n    const currentElement = this.entryRef.current\n    const calculatedDuration = this._getDuration()\n    const animationType = this._getAnimationType(calculatedDuration)\n\n    const height = isTransitionCancelled ? 'auto' : currentElement?.clientHeight ?? 'auto'\n    const transition = isTransitionCancelled\n      ? ''\n      : `height ${calculatedDuration}ms ${animationType}`\n\n    this.setState(\n      {\n        height,\n        transition,\n        looksOpenFromBreakpoint: undefined,\n      },\n      () => {\n        if (!this.props.isOpen) {\n          this._updateTransition()\n        } else {\n          this._scrollToElement(calculatedDuration)\n        }\n      }\n    )\n  }\n\n  private _getDuration() {\n    const { isOpen } = this.props\n    const currentElement = this.entryRef.current\n    const calcDuration = !!currentElement && currentElement.clientHeight / 2\n\n    if (calcDuration) {\n      if (!isOpen) {\n        return this.standardDuration\n      } else if (calcDuration > this.maxDuration) {\n        return this.maxDuration\n      } else if (calcDuration < this.standardDuration) {\n        return this.standardDuration\n      } else {\n        return calcDuration\n      }\n    } else {\n      return this.standardDuration\n    }\n  }\n\n  private _getAnimationType(calculatedDuration: number) {\n    const { isOpen } = this.props\n    return calculatedDuration === this.standardDuration ? 'linear' : isOpen ? 'ease-in' : 'ease-out'\n  }\n\n  private _updateTransition() {\n    const calculatedDuration = this._getDuration()\n    const animationType = this._getAnimationType(calculatedDuration)\n    this.setState({\n      height: this._getHeight(),\n      transition: `height ${calculatedDuration}ms ${animationType}`,\n    })\n  }\n\n  private _applyTransition() {\n    return {\n      height: this.state.height,\n      transition: this.state.transition,\n    }\n  }\n\n  private _getHeight() {\n    const { isOpen } = this.props\n    return isOpen ? 'auto' : 0\n  }\n}\n"],"file":"Foldable.js"}