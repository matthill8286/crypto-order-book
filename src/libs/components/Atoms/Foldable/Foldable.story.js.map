{"version":3,"sources":["../../../../src/components/Atoms/Foldable/Foldable.story.tsx"],"names":["withState","storiesOf","React","Foldable","toggle","store","set","checkedValue","state","module","add"],"mappings":"AAAA,SAASA,SAAT,QAA0B,0BAA1B;AACA,SAASC,SAAT,QAA0B,kBAA1B;AACA,OAAO,KAAKC,KAAZ,MAAuB,OAAvB;AACA,SAASC,QAAT;;AAEA,IAAMC,MAAM,GAAG,SAATA,MAAS,CAAAC,KAAK,EAAI;AACtBA,EAAAA,KAAK,CAACC,GAAN,CAAU;AAAEC,IAAAA,YAAY,EAAE,CAACF,KAAK,CAACG,KAAN,CAAYD;AAA7B,GAAV;AACD,CAFD;;AAIAN,SAAS,CAAC,8BAAD,EAAiCQ,MAAjC,CAAT,CAAkDC,GAAlD,CACE,SADF,EAEEV,SAAS,CAAC;AAAEO,EAAAA,YAAY,EAAE;AAAhB,CAAD,EAA0B,UAAAF,KAAK;AAAA,SACtC,oBAAC,KAAD,CAAO,QAAP,QACE,oBAAC,QAAD;AAAU,IAAA,MAAM,EAAEA,KAAK,CAACG,KAAN,CAAYD;AAA9B,cADF,EAEE;AAAQ,IAAA,OAAO,EAAE;AAAA,aAAMH,MAAM,CAACC,KAAD,CAAZ;AAAA;AAAjB,sBAFF,CADsC;AAAA,CAA/B,CAFX","sourcesContent":["import { withState } from '@dump247/storybook-state'\nimport { storiesOf } from '@storybook/react'\nimport * as React from 'react'\nimport { Foldable } from './Foldable'\n\nconst toggle = store => {\n  store.set({ checkedValue: !store.state.checkedValue })\n}\n\nstoriesOf('Design System/Atoms/Foldable', module).add(\n  'Default',\n  withState({ checkedValue: false }, store => (\n    <React.Fragment>\n      <Foldable isOpen={store.state.checkedValue}>HIDDEN</Foldable>\n      <button onClick={() => toggle(store)}>Toggle Content</button>\n    </React.Fragment>\n  ))\n)\n"],"file":"Foldable.story.js"}