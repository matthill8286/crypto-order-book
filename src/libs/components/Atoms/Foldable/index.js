export * from "./Foldable";
//# sourceMappingURL=index.js.map