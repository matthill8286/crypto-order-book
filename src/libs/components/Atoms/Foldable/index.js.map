{"version":3,"sources":["../../../../src/components/Atoms/Foldable/index.ts"],"names":[],"mappings":"AAAA","sourcesContent":["export * from './Foldable'\n"],"file":"index.js"}