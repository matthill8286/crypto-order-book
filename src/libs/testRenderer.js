import { mount } from 'enzyme';
import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { saiyanTheme } from "./styles/sc-vars-saiyan";
export function renderWithTheme(tree) {
  var theme = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : saiyanTheme;
  return renderer.create(React.createElement(ThemeProvider, {
    theme: theme
  }, tree)).toJSON();
}
export function renderWithThemeAndRouter(tree) {
  var theme = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : saiyanTheme;
  return renderer.create(React.createElement(ThemeProvider, {
    theme: theme
  }, React.createElement(Router, null, tree))).toJSON();
}
export function renderWithRouter(tree) {
  return renderer.create(React.createElement(Router, null, tree)).toJSON();
}
export function mountWithTheme(tree) {
  return mount(React.createElement(ThemeProvider, {
    theme: saiyanTheme
  }, tree));
}
export function mountWithThemeAndRouter(tree) {
  return mount(React.createElement(ThemeProvider, {
    theme: saiyanTheme
  }, React.createElement(Router, null, tree)));
}
//# sourceMappingURL=testRenderer.js.map