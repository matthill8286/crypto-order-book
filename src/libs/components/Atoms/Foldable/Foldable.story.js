import { withState } from '@dump247/storybook-state';
import { storiesOf } from '@storybook/react';
import * as React from 'react';
import { Foldable } from "./Foldable";

var toggle = function toggle(store) {
  store.set({
    checkedValue: !store.state.checkedValue
  });
};

storiesOf('Design System/Atoms/Foldable', module).add('Default', withState({
  checkedValue: false
}, function (store) {
  return React.createElement(React.Fragment, null, React.createElement(Foldable, {
    isOpen: store.state.checkedValue
  }, "HIDDEN"), React.createElement("button", {
    onClick: function onClick() {
      return toggle(store);
    }
  }, "Toggle Content"));
}));
//# sourceMappingURL=Foldable.story.js.map