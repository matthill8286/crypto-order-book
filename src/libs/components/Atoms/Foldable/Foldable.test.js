import { mount } from 'enzyme';
import React from 'react';
import { renderWithThemeAndRouter } from "../../../testRenderer";
import { Foldable, StyledFoldable } from "./Foldable";

var assertHeightIsCorrect = function assertHeightIsCorrect(wrapper, height) {
  expect(wrapper.find(StyledFoldable).prop('style').height).toBe(height);
};

describe('Foldable', function () {
  beforeAll(function () {
    Object.defineProperty(HTMLElement.prototype, 'clientHeight', {
      configurable: true,
      value: 50
    });
  });
  it('should render', function () {
    var tree = renderWithThemeAndRouter(React.createElement(Foldable, {
      isOpen: true
    }, React.createElement("div", null, "HIDDEN")));
    expect(tree).toMatchSnapshot();
  });
  it('should have a height of auto if initially opened', function () {
    var props = {
      isOpen: true
    };
    var wrapper = mount(React.createElement(Foldable, props, React.createElement("div", null, "HIDDEN")));
    expect(wrapper.state('height')).toBe(50);
    assertHeightIsCorrect(wrapper, 50);
  });
  it('should have a height of 0 if initially closed', function () {
    var props = {
      isOpen: false
    };
    var wrapper = mount(React.createElement(Foldable, props, React.createElement("div", null, "HIDDEN")));
    expect(wrapper.state('height')).toBe(0);
    assertHeightIsCorrect(wrapper, 0);
  });
  it('should update the height two times when opened', function () {
    var props = {
      isOpen: false
    };
    var wrapper = mount(React.createElement(Foldable, props, React.createElement("div", null, "HIDDEN")));
    var node = wrapper.find('[data-test="inner-foldable"]').instance();
    Object.defineProperty(node, 'clientHeight', {
      value: 100
    });
    var spy = jest.spyOn(wrapper.instance(), 'setState');
    expect(wrapper.state('height')).toBe(0);
    assertHeightIsCorrect(wrapper, 0);
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.setProps({
      isOpen: true
    });
    wrapper.childAt(0).simulate('transitionEnd');
    expect(spy).toHaveBeenCalledTimes(2);
    var calls = spy.mock.calls;
    var call1Arg1 = calls && calls[0] && calls[0][0];
    var call2Arg1 = calls && calls[1] && calls[1][0];

    if (call1Arg1 && 'height' in call1Arg1) {
      expect(call1Arg1.height).toBe(100);
    }

    if (call2Arg1 && 'height' in call2Arg1) {
      expect(call2Arg1.height).toBe('auto');
    }

    spy.mockClear();
  });
  it('should update the height two times when closed', function () {
    var props = {
      isOpen: true
    };
    var wrapper = mount(React.createElement(Foldable, props, React.createElement("div", null, "HIDDEN")));
    var node = wrapper.find('[data-test="inner-foldable"]').instance();
    Object.defineProperty(node, 'clientHeight', {
      value: 100
    });
    var spy = jest.spyOn(wrapper.instance(), 'setState');
    expect(wrapper.state('height')).toBe(50);
    assertHeightIsCorrect(wrapper, 50);
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.setProps({
      isOpen: false
    });
    expect(spy).toHaveBeenCalledTimes(2);
    var calls = spy.mock.calls;
    var call1Arg1 = calls && calls[0] && calls[0][0];
    var call2Arg1 = calls && calls[1] && calls[1][0];

    if (call1Arg1 && 'height' in call1Arg1) {
      expect(call1Arg1.height).toBe(100);
    }

    if (call2Arg1 && 'height' in call2Arg1) {
      expect(call2Arg1.height).toBe(0);
    }

    spy.mockClear();
  });
});
//# sourceMappingURL=Foldable.test.js.map