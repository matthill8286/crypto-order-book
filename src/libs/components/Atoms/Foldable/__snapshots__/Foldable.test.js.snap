// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Foldable should render 1`] = `
<div
  className="Foldable__StyledFoldable-sc-1ors3ot-0 ijHRpZ"
  data-test="foldable"
  onTransitionEnd={[Function]}
  style={
    Object {
      "height": "auto",
      "transition": "height 300ms linear",
    }
  }
>
  <div
    data-test="inner-foldable"
  >
    <div>
      HIDDEN
    </div>
  </div>
</div>
`;
